### PR TITLE
Help / Info Popup Improvements

### PR DESCRIPTION
**Pull request by lalitpatel**
### Summary
Changed a bunch of things in the Help / Info Popup to improve readability:

1. Grouped various stats into Stats, Attacks and Armours.
2. Added some margin between sections
3. Hide advanced stats by default and show it when the user clicks 'More Info' link
4. Increased the size of civ images in the info popup to use up the available width
5. Added a drop shadow to the info popup instead of a border

### Before
before

### After
one
two

### CODE REVIEW

File: index.html [hunk 1]
```diff
@@ -80,14 +80,18 @@
         <div id="techtree">
             <div id="helptext">
                 <div id="helptext__content"></div>
+                <div id="helptext__advance_stats"></div>
                 <div id="helptext__x_ref"></div>
+                <div id="helptext__more">
```
Comment (paulirish) [inline index.html:85]: you could use `<details><summary>More info</summary> .... </details>` here instead and you'd get to drop most of the `toggle_more_stats` JS

File: index.html [hunk 2]
```diff
@@ -596,6 +655,20 @@
...
     }
 
+    function toggle_more_stats() {
+        const stats = document.getElementById('helptext__advance_stats');
+        const civs = document.getElementById('helptext__x_ref');
+        const link = document.getElementById('helptext__more_link');
+        if (civs.style.display === 'none') {
+            stats.style.display = 'block';
+            civs.style.display = 'block';
+            link.innerHTML = 'Less Info &#x25B2;';
```
Comment (paulirish) [inline index.html:665]:
```suggestion
            link.textContent = 'Less Info ▲';
```

just fwiw: textContent is preferred to reduce likelihood of XSS.

File: index.html [hunk 3]
```diff
@@ -370,43 +380,87 @@
...
+            }
+            if (text.match(/‹(P|p)iercearmor›/)) {
+                stats.push("Pierce armor:&nbsp;" + meta.PierceArmor);
+            }
+            if (text.match(/‹garrison›/)) {
+                stats.push("Garrison:&nbsp;" + meta.GarrisonCapacity);
+            }
+            if (text.match(/‹range›/)) {
+                stats.push("Range:&nbsp;" + meta.Range);
+            }
```
[the remark this answers is not in the document]
Comment (lalitpatel) [reply index.html:424]: I think this was intentionally done so in the previous code. There may be certain unit's or buildings where certain stats would be better left out eg. Attack, Range, Garrison etc. For example: For a University we need not specify Garrison or Range.

File: index.html [hunk 4]
```diff
@@ -238,7 +244,9 @@
         const selectedCiv = document.getElementById('civselect').value;
         civ(selectedCiv, tree);
         if (selectedCiv in data.civ_helptexts) {
-            document.getElementById('civtext').innerHTML = data.strings[data.civ_helptexts[selectedCiv]];
+            let civText = data.strings[data.civ_helptexts[selectedCiv]];
+            civText = civText.replace(/(.*civilization\s*<br>\s*<br>)/m, "$1<b>Civilization Bonus:</b><br>")
```
[the remark this answers is not in the document]
Comment (lalitpatel) [reply index.html:248]: I added to just to make it look symmetric. Without the heading, it was looking odd to me. I do not feel strongly about it and am okay to remove it.

**Comment by lalitpatel** [comment]
> People on resolutions lower than 1080p should still be able to display the full information in the popover. While the layout you are proposing is structured a lot better visually, it is also a lot taller in comparison:

This is a very valid concern. However, I have a slightly different view point. Whenever the height of info popup is more than the height of the viewport, the popup is not cropped and it is possible for user to scroll vertically to view the rest of the popup. IMO, that is an acceptable tradeoff for vast majority of the cases where this will result in better UX. Also we should note that this is only an issue when the "Advanced Stats" section is expanded.

To reproduce the above scroll behaviour on 1080p monitors, click on the Cavalry Archer icon to display the popup and then click on Advanced Stats link. You will notice that part of the popup is outside of the viewport, however you will be able to scroll down to see the whole of the popup.

**Comment by lalitpatel** [comment]
@HSZemi Here is screenshot of a version with the changes that you suggested:

image

**Comment by HSZemi** [comment]
Even the cavalry archer helptext fully fits on my screen :shrug:
I do prefer the left, shorter option with the inline heading though.

File: index.html [hunk 5]
```diff
@@ -297,12 +310,14 @@
     function displayHelp(caretId) {
         focusedNodeId = caretId;
         let helptextContent = document.getElementById("helptext__content");
+        let helptextAdvanceStats = document.getElementById("helptext__advance_stats");
```
Comment (HSZemi) [inline index.html:313]:
```suggestion
        let helptextAdvancedStats = document.getElementById("helptext__advanced_stats");
```

File: index.html (hunk 4 above)
[the remark this answers is not in the document]
Comment (HSZemi) [reply index.html:253]: Let's remove it then

File: index.html [hunk 6]
```diff
@@ -60,34 +60,43 @@
...
-                </p>
-                <p>
-                    Age of Empires II &copy; Microsoft Corporation.</br>
-                    <b>aoe2techtree</b> was created under Microsoft's "<a href="https://www.xbox.com/en-us/developers/rules">Game Content Usage Rules</a>" using assets
+                    Age of Empires II &copy; Microsoft Corporation.<br>
+                    <b>aoe2techtree</b> was created under Microsoft's "<a href="https://www.xbox.com/en-us/developers/rules" rel="nofollow">Game Content Usage Rules</a>" using assets
                     from Age of Empires II, and it is not endorsed by or affiliated with Microsoft.
                 </p>
+                <details>
+                    <summary>Credits</summary>
```
Comment (HSZemi) [inline index.html:71]: Do we want to "hide" the credits like this? :thinking:

Comment (lalitpatel) [reply index.html:71]: My intention was to reduce the text on that side bar 😄

Comment (lalitpatel) [reply index.html:71]: Also, we may need some space for the new feature that I was planning to submit a PR for: Color Key / Legend
 
image

Comment (HSZemi) [reply index.html:71]: ayayay

Comment (lalitpatel) [reply index.html:71]: So what is the final decision on it? This is fine or should I revert back the credits section to how it was earlier?

> Version: Age of Empires II DE Update 36906
> 
> Contribute on GitHub
> 
> Made by hszemi, Anda, exterkamp, paulirish,
> with thanks to Jineapple, TriRem, pip, and NkoDragaš
> Item Metadata (cost, HP etc.) taken from aoe2dat
> 
> Age of Empires II © Microsoft Corporation.
> aoe2techtree was created under Microsoft's "Game Content Usage Rules" using assets from Age of Empires II, and it is not endorsed by or affiliated with Microsoft.

Comment (HSZemi) [reply index.html:71]: Let's keep the old version for now and revisit this once the legend is suggested

Comment (lalitpatel) [reply index.html]: Done!

File: style.css [hunk 7]
```diff
@@ -53,6 +53,7 @@ a, a:hover, a:visited, a:active {
     border: 2px solid #bbb;
     padding: 0.4rem;
     font-size: 10pt;
+    margin:5px;
```
Comment (HSZemi) [inline style.css:56]: This seems to negatively affect the positioning, for example the corners do not touch anymore
grafik

Comment (lalitpatel) [reply style.css]: Oh 🤦 
Thanks for QA-ing this so well. 👍

**Comment by HSZemi** [comment]
Looks like it's time to merge!

File: index.html [hunk 8]
```diff
@@ -370,43 +378,87 @@
...
+            }
+            stats.push(minRangeIfDefined(meta.MinRange, "Min Range:&nbsp;"));
+            stats.push(ifDefined(meta.LineOfSight, "Line of Sight:&nbsp;"));
+            stats.push(ifDefined(meta.Speed, "Speed:&nbsp;"));
+            stats.push(secondsIfDefined(meta.TrainTime, "Build Time:&nbsp;"));
+            stats.push(secondsIfDefined(meta.ResearchTime, "Research Time:&nbsp;"));
+            stats.push(ifDefined(meta.FrameDelay, "Frame Delay:&nbsp;"));
+            stats.push(secondsIfDefined(meta.ReloadTime, "Reload Time:&nbsp;"));
+            stats.push(accuracyIfDefined(meta.AccuracyPercent, "Accuracy:&nbsp;"));
+            text = text.replace(/<p class="helptext__stats">(.+?)<\/p>/, "<h3>Stats</h3><p>" + stats.filter(Boolean).join(', ') + "<p>")
```
Comment (pardoman) [inline index.html:431]: This class in only used within this function; is it supposed to have some css styling associated to it?

Comment (lalitpatel) [reply index.html:431]: This helps identify the `<p>` for stats which I want to replace since there are multiple `<p>` in the text.

**Pull request merged**